tests/replacer/test_single_nodes.py

```diff
@@ -0,0 +1,342 @@
+            (ClickhouseNode("storage-2-1", 9000, 3, 2), healthy),
+        ],
+    )
+
+
+@pytest.fixture
+def override_cluster() -> Generator[
+    Callable[[bool], FakeClickhouseCluster], None, None
+]:
+    current_clusters = cluster.CLUSTERS
+    current_mapping = cluster._STORAGE_SET_CLUSTER_MAP
+
+    def override(healthy: bool) -> FakeClickhouseCluster:
+        test_cluster = _build_cluster(healthy=healthy)
+        cluster.CLUSTERS = [test_cluster]
+        cluster._STORAGE_SET_CLUSTER_MAP = {StorageSetKey.EVENTS: cluster.CLUSTERS[0]}
+        return test_cluster
+
+    try:
+        yield override
+    finally:
+        cluster.CLUSTERS = current_clusters
+        cluster._STORAGE_SET_CLUSTER_MAP = current_mapping
+
+
+LOCAL_QUERY = """\
+INSERT INTO errors_local (project_id, timestamp, event_id)
+SELECT project_id, timestamp, event_id, group_id, primary_hash
+FROM errors_local FINAL
+WHERE event_id = '6f0ccc03-6efb-4f7c-8005-d0c992106b31'
+"""
+
+DIST_QUERY = """\
+INSERT INTO errors_dist (project_id, timestamp, event_id)
+SELECT project_id, timestamp, event_id, group_id, primary_hash
+FROM errors_dist FINAL
+WHERE event_id = '6f0ccc03-6efb-4f7c-8005-d0c992106b31'
+"""
+
+TEST_CASES = [
+    pytest.param(
+        "override_cluster",
+        "[100,1]",
+        {
+            "query_node": [
+                "SELECT count() FROM errors_dist FINAL WHERE event_id = '6f0ccc03-6efb-4f7c-8005-d0c992106b31'",
+            ],
+            "storage-0-0": [LOCAL_QUERY],
+            "storage-1-0": [LOCAL_QUERY],
+            "storage-2-0": [LOCAL_QUERY],
+        },
+        id="Replacements through storage nodes",
+    ),
+    pytest.param(
+        "override_cluster",
+        "",
+        {
+            "query_node": [
+                "SELECT count() FROM errors_dist FINAL WHERE event_id = '6f0ccc03-6efb-4f7c-8005-d0c992106b31'",
+                DIST_QUERY,
+            ]
+        },
+        id="Replacements through query node",
+    ),
+    pytest.param(
+        "override_cluster",
+        "[10]",
+        {
+            "query_node": [
+                "SELECT count() FROM errors_dist FINAL WHERE event_id = '6f0ccc03-6efb-4f7c-8005-d0c992106b31'",
+                DIST_QUERY,
+            ]
+        },
+        id="Replacements through query node. Wrong project",
+    ),
+]
+
+COUNT_QUERY_TEMPLATE = "SELECT count() FROM %(table_name)s FINAL WHERE event_id = '6f0ccc03-6efb-4f7c-8005-d0c992106b31'"
+INSERT_QUERY_TEMPLATE = """\
+INSERT INTO %(table_name)s (%(required_columns)s)
+SELECT %(select_columns)s
+FROM %(table_name)s FINAL
+WHERE event_id = '6f0ccc03-6efb-4f7c-8005-d0c992106b31'
+"""
+
+FINAL_QUERY_TEMPLATE = {
+    "required_columns": "project_id, timestamp, event_id",
+    "select_columns": "project_id, timestamp, event_id, group_id, primary_hash",
+}
+
+
+@pytest.mark.parametrize(
+    "override_fixture, write_node_replacements_projects, expected_queries", TEST_CASES
+)
+def test_write_single_node(
```

Comment:
How come this refers to single nodes? I get the impression that TEST_CASES includes both kinds of scenarios.